[JSC] optimize `Number.isFinite()` by rewriting in C++

Gumichocopengin8 opened the pull request:
#### 832acac1268134ad49cd9b6a2c333c86d00234e0
<pre>
[JSC] optimize `Number.isFinite()` by rewriting in C++
<a href="https://bugs.webkit.org/show_bug.cgi?id=290630">https://bugs.webkit.org/show_bug.cgi?id=290630</a>

Reviewed by Yusuke Suzuki.

Optimized `Number.isFinite` by rewriting it in C++

                    TipOfTree              Patched                Ratio
number-isFinite     1.0720+-0.1239         0.6741+-0.1325         1.5902x faster

* JSTests/microbenchmarks/number-isFinite.js: Added.
* Source/JavaScriptCore/CMakeLists.txt:
* Source/JavaScriptCore/DerivedSources-input.xcfilelist:
* Source/JavaScriptCore/DerivedSources.make:
* Source/JavaScriptCore/JavaScriptCore.xcodeproj/project.pbxproj:
* Source/JavaScriptCore/builtins/NumberConstructor.js: Removed.
* Source/JavaScriptCore/dfg/DFGAbstractInterpreterInlines.h:
(JSC::DFG::AbstractInterpreter&lt;AbstractStateType&gt;::executeEffects):
* Source/JavaScriptCore/dfg/DFGByteCodeParser.cpp:
(JSC::DFG::ByteCodeParser::handleIntrinsicCall):
* Source/JavaScriptCore/dfg/DFGClobberize.h:
(JSC::DFG::clobberize):
* Source/JavaScriptCore/dfg/DFGDoesGC.cpp:
(JSC::DFG::doesGC):
* Source/JavaScriptCore/dfg/DFGFixupPhase.cpp:
(JSC::DFG::FixupPhase::fixupNode):
* Source/JavaScriptCore/dfg/DFGNodeType.h:
* Source/JavaScriptCore/dfg/DFGOperations.cpp:
(JSC::DFG::JSC_DEFINE_NOEXCEPT_JIT_OPERATION):
* Source/JavaScriptCore/dfg/DFGOperations.h:
* Source/JavaScriptCore/dfg/DFGPredictionPropagationPhase.cpp:
* Source/JavaScriptCore/dfg/DFGSafeToExecute.h:
(JSC::DFG::safeToExecute):
* Source/JavaScriptCore/dfg/DFGSpeculativeJIT.cpp:
* Source/JavaScriptCore/dfg/DFGSpeculativeJIT.h:
* Source/JavaScriptCore/dfg/DFGSpeculativeJIT32_64.cpp:
(JSC::DFG::SpeculativeJIT::compile):
* Source/JavaScriptCore/dfg/DFGSpeculativeJIT64.cpp:
(JSC::DFG::SpeculativeJIT::compile):
* Source/JavaScriptCore/dfg/DFGStrengthReductionPhase.cpp:
(JSC::DFG::StrengthReductionPhase::handleNode):
* Source/JavaScriptCore/ftl/FTLCapabilities.cpp:
(JSC::FTL::canCompile):
* Source/JavaScriptCore/ftl/FTLLowerDFGToB3.cpp:
(JSC::FTL::DFG::LowerDFGToB3::compileNode):
(JSC::FTL::DFG::LowerDFGToB3::compileCompareStrictEq):
* Source/JavaScriptCore/jit/ThunkGenerators.cpp:
(JSC::numberIsFiniteThunkGenerator):
* Source/JavaScriptCore/jit/ThunkGenerators.h:
* Source/JavaScriptCore/runtime/Intrinsic.h:
* Source/JavaScriptCore/runtime/NumberConstructor.cpp:
(JSC::JSC_DEFINE_HOST_FUNCTION):
* Source/JavaScriptCore/runtime/VM.cpp:
(JSC::thunkGeneratorForIntrinsic):

Canonical link: <a href="https://commits.webkit.org/293031@main">https://commits.webkit.org/293031@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/edc6c7d46e12096991dd2aec6d05f33929c15faf

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/97698 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/17323 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/7540 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/102804 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/48227 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/17617 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/25776 "Built successfully") | [❌ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/74414 "Found 1 new test failure: fast/css/view-transitions-nested-transparency-layers.html (failure)") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/31600 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/100701 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/13349 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/88316 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/54763 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/13119 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/6230 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/47670 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/90374 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/83125 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/6307 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/104806 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/96321 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/24778 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/18061 "Found 1 new test failure: ipc/send-gpu-GetShareableBitmap-RemoteRenderingBackend.html (failure)") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/83464 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/25150 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/84474 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/82897 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/27463 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/5141 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/18378 "The change is no longer eligible for processing.") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/15803 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/24739 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/29908 "Built successfully") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/119946 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/24561 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/33665 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/27875 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/26135 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

Constellation reviewed (changes requested):
This function is intentionally written in JS to allow the caller to inline it and examine the effect of this. When just comparing the numbers with this one function call, C++ can be faster, but it is not when it is used in the larger function and this examined effect allows the other optimizations around it. So if we would like to put it to C++, you need to implement DFG / FTL handling too. See `NumberIsNaNIntrinsic` handling in DFG / FTL for example.

Constellation commented on Source/JavaScriptCore/dfg/DFGSpeculativeJIT.cpp:
```diff
@@ -16748,6 +16748,52 @@ void SpeculativeJIT::compileNumberIsNaN(Node* node)
+        FPRTemporary diff(this);
+        FPRReg diffFPR = diff.fpr();
+        FPRTemporary neg(this);
+        FPRReg negFPR = neg.fpr();
```
Let's follow to the typical procedure, first defining all operands and temporaries, and then getting gpr() etc.
```
SpeculateDoubleOperand argument(this, node->child1());
GPRTemporary scratch(this);
FPRTemporary diff(this);
FPRTemporary neg(this);

FPRReg argumentFPR = argument.fpr();
GPRReg scratchGPR = scratch.gpr();
FPRReg diffFPR = diff.fpr();
FPRReg negFPR = neg.fpr();
```

Gumichocopengin8 replied:
Thanks. Fixed. Is it preferable to use`add + neg` over `sub`?

Constellation commented on Source/JavaScriptCore/dfg/DFGSpeculativeJIT.cpp:
```diff
@@ -16748,6 +16748,52 @@ void SpeculativeJIT::compileNumberIsNaN(Node* node)
+        FPRTemporary neg(this);
+        FPRReg negFPR = neg.fpr();
+
+        negateDouble(negFPR, argumentFPR);
```
This looks not correct. Isn't it opposite?

Constellation commented on Source/JavaScriptCore/dfg/DFGSpeculativeJIT.cpp:
```diff
@@ -16748,6 +16748,52 @@ void SpeculativeJIT::compileNumberIsNaN(Node* node)
+        FPRReg negFPR = neg.fpr();
+
+        negateDouble(negFPR, argumentFPR);
+        addDouble(diffFPR, argumentFPR, negFPR);
```
Ditto.

Constellation commented on Source/JavaScriptCore/ftl/FTLCapabilities.cpp:
```diff
@@ -641,4 +642,3 @@ CapabilityLevel canCompile(Graph& graph)
 } } // namespace JSC::FTL
 
 #endif // ENABLE(FTL_JIT)
-
```
Remove this change.

Constellation commented on Source/JavaScriptCore/dfg/DFGStrengthReductionPhase.cpp:
```diff
@@ -1756,4 +1757,3 @@ bool performStrengthReduction(Graph& graph)
 } } // namespace JSC::DFG
 
 #endif // ENABLE(DFG_JIT)
-
```
Remove this change.

Constellation commented on Source/JavaScriptCore/dfg/DFGPredictionPropagationPhase.cpp:
```diff
@@ -1803,4 +1804,3 @@ bool performPredictionPropagation(Graph& graph)
 } } // namespace JSC::DFG
 
 #endif // ENABLE(DFG_JIT)
-
```
Remove this change.

Constellation commented on Source/JavaScriptCore/runtime/Intrinsic.h:
```diff
@@ -257,5 +258,3 @@ class PrintStream;
 
 } // namespace WTF
-
-
```
Remove this change.

Constellation commented on Source/JavaScriptCore/runtime/NumberConstructor.cpp:
```diff
@@ -161,4 +162,11 @@ JSC_DEFINE_HOST_FUNCTION(numberConstructorFuncIsNaN, (JSGlobalObject*, CallFrame
+    if (!argument.isNumber())
+        return JSValue::encode(jsBoolean(false));
+    return JSValue::encode(jsBoolean(std::isfinite(argument.asNumber())));
+}
```
Add empty line after this `}`.

Constellation commented on Source/JavaScriptCore/ftl/FTLLowerDFGToB3.cpp:
```diff
@@ -13971,6 +13974,40 @@ IGNORE_CLANG_WARNINGS_END
+        switch (m_node->child1().useKind()) {
+        case DoubleRepUse: {
+            LValue argument = lowDouble(m_node->child1());
+            setBoolean(vmCall(Int32, operationNumberIsFinite, argument));
```
We should have inlined code aligned to DFG.

Constellation commented on Source/JavaScriptCore/dfg/DFGSpeculativeJIT.cpp:
```diff
@@ -16748,6 +16748,53 @@ void SpeculativeJIT::compileNumberIsNaN(Node* node)
+        FPRReg negFPR = neg.fpr();
+
+        negateDouble(argumentFPR, negFPR);
+        addDouble(argumentFPR, negFPR, diffFPR);
```
Yeah, if we can just use `subDouble` here, that's better.
```
subDouble(argumentFPR, argumentFPR, diffFPR);
```

Constellation replied:
And also can you add a test? It seems that they are missed in the test.

Gumichocopengin8 replied:
Replaced it with `subDouble`. Added `LayoutTests/js/script-tests/Number-isFinite.js`

Constellation commented on Source/JavaScriptCore/ftl/FTLLowerDFGToB3.cpp:
```diff
@@ -13971,6 +13974,42 @@ IGNORE_CLANG_WARNINGS_END
+        case DoubleRepUse: {
+            LValue argument = lowDouble(m_node->child1());
+            LValue negArgument = m_out.neg(argument);
+            LValue result = m_out.doubleAdd(argument, negArgument);
```
It is worth using `m_out.doubleSub`.

Gumichocopengin8 replied:
Replaced

Constellation commented on Source/JavaScriptCore/ftl/FTLLowerDFGToB3.cpp:
```diff
@@ -13971,6 +13974,42 @@ IGNORE_CLANG_WARNINGS_END
+            LValue argument = lowDouble(m_node->child1());
+            LValue negArgument = m_out.neg(argument);
+            LValue result = m_out.doubleAdd(argument, negArgument);
+            setBoolean(m_out.equal(result, result));
```
Is it correct? (It is `equal`, not `doubleEqual`). Also can you add tests for them?

Gumichocopengin8 replied:
Fixed

Constellation commented on JSTests/microbenchmarks/number-isFinite.js:
```diff
@@ -0,0 +1,20 @@
+  ];
+  for (var i = 0; i < testLoopCount; ++i) {
+    for (var j = 0; j < values.length; ++j) {
+      Number.isFinite(values[j]);
```
This is not checking the result.

Constellation reviewed (approved):
Looks good! r=me

Gumichocopengin8 committed:
[JSC] optimize `Number.isFinite()` by rewriting in C++
https://bugs.webkit.org/show_bug.cgi?id=290630

Reviewed by Yusuke Suzuki.

Optimized `Number.isFinite` by rewriting it in C++

                    TipOfTree              Patched                Ratio
number-isFinite     1.0720+-0.1239         0.6741+-0.1325         1.5902x faster

* JSTests/microbenchmarks/number-isFinite.js: Added.
* Source/JavaScriptCore/CMakeLists.txt:
* Source/JavaScriptCore/DerivedSources-input.xcfilelist:
* Source/JavaScriptCore/DerivedSources.make:
* Source/JavaScriptCore/JavaScriptCore.xcodeproj/project.pbxproj:
* Source/JavaScriptCore/builtins/NumberConstructor.js: Removed.
* Source/JavaScriptCore/dfg/DFGAbstractInterpreterInlines.h:
(JSC::DFG::AbstractInterpreter<AbstractStateType>::executeEffects):
* Source/JavaScriptCore/dfg/DFGByteCodeParser.cpp:
(JSC::DFG::ByteCodeParser::handleIntrinsicCall):
* Source/JavaScriptCore/dfg/DFGClobberize.h:
(JSC::DFG::clobberize):
* Source/JavaScriptCore/dfg/DFGDoesGC.cpp:
(JSC::DFG::doesGC):
* Source/JavaScriptCore/dfg/DFGFixupPhase.cpp:
(JSC::DFG::FixupPhase::fixupNode):
* Source/JavaScriptCore/dfg/DFGNodeType.h:
* Source/JavaScriptCore/dfg/DFGOperations.cpp:
(JSC::DFG::JSC_DEFINE_NOEXCEPT_JIT_OPERATION):
* Source/JavaScriptCore/dfg/DFGOperations.h:
* Source/JavaScriptCore/dfg/DFGPredictionPropagationPhase.cpp:
* Source/JavaScriptCore/dfg/DFGSafeToExecute.h:
(JSC::DFG::safeToExecute):
* Source/JavaScriptCore/dfg/DFGSpeculativeJIT.cpp:
* Source/JavaScriptCore/dfg/DFGSpeculativeJIT.h:
* Source/JavaScriptCore/dfg/DFGSpeculativeJIT32_64.cpp:
(JSC::DFG::SpeculativeJIT::compile):
* Source/JavaScriptCore/dfg/DFGSpeculativeJIT64.cpp:
(JSC::DFG::SpeculativeJIT::compile):
* Source/JavaScriptCore/dfg/DFGStrengthReductionPhase.cpp:
(JSC::DFG::StrengthReductionPhase::handleNode):
* Source/JavaScriptCore/ftl/FTLCapabilities.cpp:
(JSC::FTL::canCompile):
* Source/JavaScriptCore/ftl/FTLLowerDFGToB3.cpp:
(JSC::FTL::DFG::LowerDFGToB3::compileNode):
(JSC::FTL::DFG::LowerDFGToB3::compileCompareStrictEq):
* Source/JavaScriptCore/jit/ThunkGenerators.cpp:
(JSC::numberIsFiniteThunkGenerator):
* Source/JavaScriptCore/jit/ThunkGenerators.h:
* Source/JavaScriptCore/runtime/Intrinsic.h:
* Source/JavaScriptCore/runtime/NumberConstructor.cpp:
(JSC::JSC_DEFINE_HOST_FUNCTION):
* Source/JavaScriptCore/runtime/VM.cpp:
(JSC::thunkGeneratorForIntrinsic):

Canonical link: https://commits.webkit.org/293031@main
